File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    static List<int[]> dungeonList = new ArrayList<>();
    static List<Integer> results = new ArrayList<>();
    
    public int solution(int k, int[][] dungeons) {
        for(int[] dungeon: dungeons) {
            dungeonList.add(dungeon);
        }
        
        dfs(dungeonList, k, 0);
        
        return Collections.max(results);
    }
    
    private void dfs(List<int[]> dungeonList, int k, int count) {
        if(dungeonList.size() == 0) {
            results.add(count);
            return;
        }
        
        for(int i = 0; i < dungeonList.size(); i++) {
            int[] dungeon = dungeonList.get(i);
            
            List<int[]> newDungeons = new ArrayList<>();
            for(int j = 0; j < dungeonList.size(); j++) {
                if(i != j) {
                    newDungeons.add(dungeonList.get(j));
                }
            }
            
            if(k >= dungeon[0]) {
                dfs(newDungeons, k - dungeon[1], count + 1);
            } else {
                dfs(newDungeons, k, count);
            }
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = -1
    
    N = len(dungeons)
    
    # p: 남은 피로도
    def dfs(p, cnt):
        nonlocal answer
        answer = max(answer, cnt)
        
        for i in range(N):
            if visited[i]:
                continue

            # cp: 최소 필요 피로도, sp: 소모 피로도
            cp, sp = dungeons[i]
            if cp > p:
                continue;

            visited[i] = True
            dfs(p - sp, cnt + 1)
            visited[i] = False
            
    for i in range(N):
        visited = [False] * N
        dfs(k, 0)
    
    return answer